### ISSUE TITLE
linuxkit run failing recently with d4m edge/stable on high sierra and mojave

### ISSUE THREAD

**Issue by dweomer** [linuxkit/linuxkit#3289]
<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/linuxkit/linuxkit/blob/master/CONTRIBUTING.md#reporting-other-issues

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

`linuxkit run` fails with "cannot run hyperkit: signal: abort trap". This seems to happen with hyperkit/vpnkit installed with Docker for Mac (both edge and stable) on High Sierra and Mojave.

**Steps to reproduce the issue:**

1. [Install Docker for Mac](https://hub.docker.com/editions/community/docker-ce-desktop-mac)
2. [Install LinuxKit](https://github.com/linuxkit/linuxkit#getting-started)
  a. `brew tap linuxkit/linuxkit`
  b. `brew install --HEAD linuxkit`
3. Build and run the [minimal example](https://github.com/linuxkit/linuxkit/blob/master/examples/minimal.yml)
  a. `linuxkit build -dir $(mkdir -p linuxkit;echo linuxkit) -format kernel+squashfs minimal.yml`
  b. `linuxkit -v run -squashfs linuxkit/minimal`

**Describe the results you received:**

```
 jblainch  jblainch-mac01  ~  Projects  local-linuxkit  $  linuxkit -v run -squashfs linuxkit/minimal
DEBU[0000] hyperkit: Start &hyperkit.HyperKit{HyperKit:"/usr/local/bin/com.docker.hyperkit", Argv0:"", StateDir:"linuxkit/minimal-state", VPNKitSock:"/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock", VPNKitUUID:"c9adb90c-4831-4235-9597-3fb5b7de0fdf", VPNKitPreferredIPv4:"", UUID:"5c35c912-1a3a-41be-87a2-194d3e88c13e", Disks:[]hyperkit.Disk{(*hyperkit.RawDisk)(0xc4200ab440)}, ISOImages:[]string(nil), VSock:true, VSockDir:"", VSockPorts:[]int{}, VSockGuestCID:3, VMNet:false, Sockets9P:[]hyperkit.Socket9P(nil), Kernel:"linuxkit/minimal-kernel", Initrd:"", Bootrom:"", CPUs:1, Memory:1024, Console:0, Pid:0, Arguments:[]string(nil), CmdLine:"", process:(*os.Process)(nil)} 
DEBU[0000] hyperkit: check &hyperkit.HyperKit{HyperKit:"/usr/local/bin/com.docker.hyperkit", Argv0:"", StateDir:"linuxkit/minimal-state", VPNKitSock:"/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock", VPNKitUUID:"c9adb90c-4831-4235-9597-3fb5b7de0fdf", VPNKitPreferredIPv4:"", UUID:"5c35c912-1a3a-41be-87a2-194d3e88c13e", Disks:[]hyperkit.Disk{(*hyperkit.RawDisk)(0xc4200ab440)}, ISOImages:[]string(nil), VSock:true, VSockDir:"", VSockPorts:[]int{}, VSockGuestCID:3, VMNet:false, Sockets9P:[]hyperkit.Socket9P(nil), Kernel:"linuxkit/minimal-kernel", Initrd:"", Bootrom:"", CPUs:1, Memory:1024, Console:0, Pid:0, Arguments:[]string(nil), CmdLine:"", process:(*os.Process)(nil)} 
ERRO[0000] hyperkit: Cannot safely shrink "linuxkit/minimal-squashfs.img" from 45MiB to 0MiB 
DEBU[0000] hyperkit: Arguments: []string{"-A", "-u", "-F", "linuxkit/minimal-state/hyperkit.pid", "-c", "1", "-m", "1024M", "-s", "0:0,hostbridge", "-s", "31,lpc", "-s", "1:0,virtio-vpnkit,path=/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock,uuid=c9adb90c-4831-4235-9597-3fb5b7de0fdf", "-U", "5c35c912-1a3a-41be-87a2-194d3e88c13e", "-s", "2:0,virtio-blk,linuxkit/minimal-squashfs.img", "-s", "3,virtio-sock,guest_cid=3,path=linuxkit/minimal-state", "-s", "4,virtio-rnd", "-l", "com1,stdio,log=linuxkit/minimal-state/console-ring", "-f", "kexec,linuxkit/minimal-kernel,,earlyprintk=serial console=tty0 console=ttyS0 console=ttyAMA0 root=/dev/vda"} 
DEBU[0000] hyperkit: CmdLine: "/usr/local/bin/com.docker.hyperkit -A -u -F linuxkit/minimal-state/hyperkit.pid -c 1 -m 1024M -s 0:0,hostbridge -s 31,lpc -s 1:0,virtio-vpnkit,path=/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock,uuid=c9adb90c-4831-4235-9597-3fb5b7de0fdf -U 5c35c912-1a3a-41be-87a2-194d3e88c13e -s 2:0,virtio-blk,linuxkit/minimal-squashfs.img -s 3,virtio-sock,guest_cid=3,path=linuxkit/minimal-state -s 4,virtio-rnd -l com1,stdio,log=linuxkit/minimal-state/console-ring -f kexec,linuxkit/minimal-kernel,,earlyprintk=serial console=tty0 console=ttyS0 console=ttyAMA0 root=/dev/vda" 
DEBU[0000] hyperkit: Starting &exec.Cmd{Path:"/usr/local/bin/com.docker.hyperkit", Args:[]string{"/usr/local/bin/com.docker.hyperkit", "-A", "-u", "-F", "linuxkit/minimal-state/hyperkit.pid", "-c", "1", "-m", "1024M", "-s", "0:0,hostbridge", "-s", "31,lpc", "-s", "1:0,virtio-vpnkit,path=/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock,uuid=c9adb90c-4831-4235-9597-3fb5b7de0fdf", "-U", "5c35c912-1a3a-41be-87a2-194d3e88c13e", "-s", "2:0,virtio-blk,linuxkit/minimal-squashfs.img", "-s", "3,virtio-sock,guest_cid=3,path=linuxkit/minimal-state", "-s", "4,virtio-rnd", "-l", "com1,stdio,log=linuxkit/minimal-state/console-ring", "-f", "kexec,linuxkit/minimal-kernel,,earlyprintk=serial console=tty0 console=ttyS0 console=ttyAMA0 root=/dev/vda"}, Env:[]string{"SHELL=/usr/local/bin/bash", "XPC_FLAGS=0x0", "no_proxy=*.local, 169.254/16, localhost, *.intel.com", "TERM_PROGRAM_VERSION=421.1", "SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.Rt3MpP8C8q/Listeners", "TERM_SESSION_ID=1D2E722B-5581-4BB1-BAB5-E92625E28679", "PWD=/Users/jblainch/Projects/local-linuxkit", "LOGNAME=jblainch", "HOME=/Users/jblainch", "LANG=en_US.UTF-8", "TMPDIR=/var/folders/7s/bwc8hz3n5498r68fw4gq6cxr0000gq/T/", "https_proxy=http://proxy-chain.intel.com:912", "TERM=xterm-256color", "USER=jblainch", "NO_PROXY=*.local, 169.254/16, localhost, *.intel.com", "DISPLAY=/private/tmp/com.apple.launchd.G2g0GBzxdm/org.macosforge.xquartz:0", "SHLVL=1", "HTTPS_PROXY=http://proxy-chain.intel.com:912", "HTTP_PROXY=http://proxy-chain.intel.com:911", "http_proxy=http://proxy-chain.intel.com:911", "XPC_SERVICE_NAME=0", "ALL_PROXY=socks5://proxy-us.intel.com:1080", "Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.WbNWPkU6Jt/Render", "all_proxy=socks5://proxy-us.intel.com:1080", "PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/Wireshark.app/Contents/MacOS", "TERM_PROGRAM=Apple_Terminal", "OLDPWD=/Users/jblainch", "_=/usr/local/bin/linuxkit"}, Dir:"", Stdin:(*os.File)(0xc4200c4000), Stdout:(*os.File)(0xc4200c4008), Stderr:(*os.File)(0xc4200c4010), ExtraFiles:[]*os.File(nil), SysProcAttr:(*syscall.SysProcAttr)(nil), Process:(*os.Process)(nil), ProcessState:(*os.ProcessState)(nil), ctx:context.Context(nil), lookPathErr:error(nil), finished:false, childFiles:[]*os.File(nil), closeAfterStart:[]io.Closer(nil), closeAfterWait:[]io.Closer(nil), goroutine:[]func() error(nil), errch:(chan error)(nil), waitDone:(chan struct {})(nil)} 
DEBU[0000] hyperkit: Pid is 14115                       
DEBU[0000] hyperkit: Waiting for &exec.Cmd{Path:"/usr/local/bin/com.docker.hyperkit", Args:[]string{"/usr/local/bin/com.docker.hyperkit", "-A", "-u", "-F", "linuxkit/minimal-state/hyperkit.pid", "-c", "1", "-m", "1024M", "-s", "0:0,hostbridge", "-s", "31,lpc", "-s", "1:0,virtio-vpnkit,path=/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock,uuid=c9adb90c-4831-4235-9597-3fb5b7de0fdf", "-U", "5c35c912-1a3a-41be-87a2-194d3e88c13e", "-s", "2:0,virtio-blk,linuxkit/minimal-squashfs.img", "-s", "3,virtio-sock,guest_cid=3,path=linuxkit/minimal-state", "-s", "4,virtio-rnd", "-l", "com1,stdio,log=linuxkit/minimal-state/console-ring", "-f", "kexec,linuxkit/minimal-kernel,,earlyprintk=serial console=tty0 console=ttyS0 console=ttyAMA0 root=/dev/vda"}, Env:[]string{"SHELL=/usr/local/bin/bash", "XPC_FLAGS=0x0", "no_proxy=*.local, 169.254/16, localhost, *.intel.com", "TERM_PROGRAM_VERSION=421.1", "SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.Rt3MpP8C8q/Listeners", "TERM_SESSION_ID=1D2E722B-5581-4BB1-BAB5-E92625E28679", "PWD=/Users/jblainch/Projects/local-linuxkit", "LOGNAME=jblainch", "HOME=/Users/jblainch", "LANG=en_US.UTF-8", "TMPDIR=/var/folders/7s/bwc8hz3n5498r68fw4gq6cxr0000gq/T/", "https_proxy=http://proxy-chain.intel.com:912", "TERM=xterm-256color", "USER=jblainch", "NO_PROXY=*.local, 169.254/16, localhost, *.intel.com", "DISPLAY=/private/tmp/com.apple.launchd.G2g0GBzxdm/org.macosforge.xquartz:0", "SHLVL=1", "HTTPS_PROXY=http://proxy-chain.intel.com:912", "HTTP_PROXY=http://proxy-chain.intel.com:911", "http_proxy=http://proxy-chain.intel.com:911", "XPC_SERVICE_NAME=0", "ALL_PROXY=socks5://proxy-us.intel.com:1080", "Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.WbNWPkU6Jt/Render", "all_proxy=socks5://proxy-us.intel.com:1080", "PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/Wireshark.app/Contents/MacOS", "TERM_PROGRAM=Apple_Terminal", "OLDPWD=/Users/jblainch", "_=/usr/local/bin/linuxkit"}, Dir:"", Stdin:(*os.File)(0xc4200c4000), Stdout:(*os.File)(0xc4200c4008), Stderr:(*os.File)(0xc4200c4010), ExtraFiles:[]*os.File(nil), SysProcAttr:(*syscall.SysProcAttr)(nil), Process:(*os.Process)(0xc4203a2a20), ProcessState:(*os.ProcessState)(nil), ctx:context.Context(nil), lookPathErr:error(nil), finished:false, childFiles:[]*os.File{(*os.File)(0xc4200c4000), (*os.File)(0xc4200c4008), (*os.File)(0xc4200c4010)}, closeAfterStart:[]io.Closer(nil), closeAfterWait:[]io.Closer(nil), goroutine:[]func() error(nil), errch:(chan error)(0xc4203f22a0), waitDone:(chan struct {})(nil)} 
virtio-net-vpnkit: initialising, opts="path=/Users/jblainch/Library/Containers/com.docker.docker/Data/vpnkit.eth.sock,uuid=c9adb90c-4831-4235-9597-3fb5b7de0fdf"
virtio-net-vpnkit: interface will have uuid c9adb90c-4831-4235-9597-3fb5b7de0fdf
virtio-net-vpnkit: magic=VMN3T version=22 commit=0123456789012345678901234567890123456789??L?
virtio-net-vpnkit: Connection established with MAC=02:50:00:00:00:05 and MTU 1500
vsock init 3:0 = linuxkit/minimal-state, guest_cid = 00000003
vm exit[0]
          	reason		VMX
                                   	rip		0x000000000009e019
                                                                          	inst_length	3
                                                                                                 	status		0
                                                                                                                         	exit_reason	33
                                                                                                                                                  	qualification	0x0000000000000000
                                                                                                                                                                                          	inst_type		0
                                                                                                                                                                                                                         	inst_erro0
                                                                                                                                                                                                                                         VMCS_PIN_BASED_CTLS:           0x000000000000003f
                                                VMCS_PRI_PROC_BASED_CTLS:      0x00000000b5186dfa
                                                                                                 VMCS_SEC_PROC_BASED_CTLS:      0x00000000000000aa
                                                                                                                                                  VMCS_ENTRY_CTLS:               0x00000000000093ff
                                                                                                                                                                                                   VMCS_EXCEPTION_BITMAP:         0x0000000000040000
          VMCS_CR0_MASK:                 0x00000000e0000031
                                                           VMCS_CR0_SHADOW:               0x0000000000000001
                                                                                                            VMCS_CR4_MASK:                 0x0000000000002000
                                                                                                                                                             VMCS_CR4_SHADOW:               0x0000000000000000
                                                                                                                                                                                                              VMCS_GUEST_PHYSICAL_ADDRESS:   0x000000000009e000
                     VMCS_GUEST_LINEAR_ADDRESS:     0x000000000009e000
                                                                      VMCS_GUEST_CS_SELECTOR:        0x0000000000000008
                                                                                                                       VMCS_GUEST_CS_LIMIT:           0x00000000ffffffff
                                                                                                                                                                        VMCS_GUEST_CS_ACCESS_RIGHTS:   0x000000000000c09b
                                                                                                                                                                                                                         VMCS_GUEST_CS_BASE:            0x0000000000000000
                                VMCS_GUEST_DS_SELECTOR:        0x0000000000000018
                                                                                 VMCS_GUEST_DS_LIMIT:           0x00000000ffffffff
                                                                                                                                  VMCS_GUEST_DS_ACCESS_RIGHTS:   0x000000000000c093
                                                                                                                                                                                   VMCS_GUEST_DS_BASE:            0x0000000000000000
                                                                                                                                                                                                                                    VMCS_GUEST_ES_SELECTOR:        0x0000000000000000
                                           VMCS_GUEST_ES_LIMIT:           0x00000000ffffffff
                                                                                            VMCS_GUEST_ES_ACCESS_RIGHTS:   0x000000000001c000
                                                                                                                                             VMCS_GUEST_ES_BASE:            0x0000000000000000
                                                                                                                                                                                              VMCS_GUEST_FS_SELECTOR:        0x0000000000000000
     VMCS_GUEST_FS_LIMIT:           0x00000000ffffffff
                                                      VMCS_GUEST_FS_ACCESS_RIGHTS:   0x000000000001c000
                                                                                                       VMCS_GUEST_FS_BASE:            0x0000000000000000
                                                                                                                                                        VMCS_GUEST_GS_SELECTOR:        0x0000000000000000
                                                                                                                                                                                                         VMCS_GUEST_GS_LIMIT:           0x00000000ffffffff
                VMCS_GUEST_GS_ACCESS_RIGHTS:   0x000000000001c000
                                                                 VMCS_GUEST_GS_BASE:            0x0000000000000000
                                                                                                                  VMCS_GUEST_SS_SELECTOR:        0x0000000000000018
                                                                                                                                                                   VMCS_GUEST_SS_LIMIT:           0x00000000ffffffff
                                                                                                                                                                                                                    VMCS_GUEST_SS_ACCESS_RIGHTS:   0x000000000000c093
                           VMCS_GUEST_SS_BASE:            0x0000000000000000
                                                                            VMCS_GUEST_LDTR_SELECTOR:      0x0000000000000000
                                                                                                                             VMCS_GUEST_LDTR_LIMIT:         0x00000000ffffffff
                                                                                                                                                                              VMCS_GUEST_LDTR_ACCESS_RIGHTS: 0x000000000001c000
                                                                                                                                                                                                                               VMCS_GUEST_LDTR_BASE:          0x0000000000000000
                                      VMCS_GUEST_TR_SELECTOR:        0x0000000000000020
                                                                                       VMCS_GUEST_TR_LIMIT:           0x0000000000000fff
                                                                                                                                        VMCS_GUEST_TR_ACCESS_RIGHTS:   0x000000000000808b
                                                                                                                                                                                         VMCS_GUEST_TR_BASE:            0x0000000000000000
                                                                                                                                                                                                                                         VMCS_GUEST_GDTR_LIMIT:         0x0000000000000030
                                                VMCS_GUEST_GDTR_BASE:          0x0000000001783bf8
                                                                                                 VMCS_GUEST_IDTR_LIMIT:         0x00000000ffffffff
                                                                                                                                                  VMCS_GUEST_IDTR_BASE:          0x0000000000000000
                                                                                                                                                                                                   VMCS_GUEST_CR0:                0x0000000000000031
          VMCS_GUEST_CR3:                0x0000000002d2b000
                                                           VMCS_GUEST_CR4:                0x0000000000002020
                                                                                                            VMCS_GUEST_IA32_EFER:          0x0000000000000500

                                                                                                                                                             rip: 0x000000000009e019 rfl: 0x0000000000000047 cr2: 0x0000000000000000
                                                                                                                                                                                                                                    rax: 0x0000000000000001 rbx: 0x0000000002590000 rcx: 0x000000000009d000 rdx: 0x0000000000000000
                                                                                         rsi: 0x0000000000003000 rdi: 0x0000000001000294 rbp: 0x0000000001000000 rsp: 0x000000000009f000
                                                                                                                                                                                        r8:  0x000000000009d000 r9:  0x0000000000000000 r10: 0x0000000000000000 r11: 0x0000000000000000
                                             r12: 0x0000000000000000 r13: 0x0000000000000000 r14: 0x0000000000000000 r15: 0x0000000000000000
                                                                                                                                            FATA[0000] Cannot run hyperkit: signal: abort trap      
                                                                                                                                                                                                     jblainch  jblainch-mac01  ~  Projects  local-linuxkit  ERROR  $   jblainch  jblainch-mac01  ~  Projects  local-linuxkit  ERROR  $   jblainch  jblainch-mac01  ~  Projects  local-linuxkit  ERROR  $  
```

Additionally, this totally hoses my terminal but an `stty sane; clear` seems to fix that.

**Describe the results you expected:**

A running VM console.

**Additional information you deem important (e.g. issue happens only occasionally):**

I am able to mitigate this issue by moving the `/usr/local/bin/{hyper,vpn}kit` symlinks to `/usr/local/bin/com.docker.{hyper,vpn}kit` and dropping in the latest binaries from the hyperkit and vpnkit CircleCI builds, i.e.

- https://circleci.com/gh/moby/vpnkit/1357#artifacts/containers/0
- https://circleci.com/gh/moby/hyperkit/659#artifacts/containers/0

_Perhaps this is actually a bundling/install issue with Docker-for-Mac upstream?_

**Comment by dweomer** [comment]
I started running into this problem with recent (from late January on) Docker-for-Mac builds (2.x?)

**Comment by justincormack** [comment]
Hi, yes I noticed this today too. We can see if we can get Desktop to update their hyperkit (cc @djs55). Otherwise you can use `linuxkit run qemu` that also uses hypervisor framework on new builds.

**Comment by rn** [comment]
closing this as this is fixed in hyperkit with https://github.com/moby/hyperkit/pull/239. Don't know when Docker for Mac decides to update HyperKit.

**Comment by ijc** [comment]
FYI I just tagged Hyperkit v0.20190201 since that is the version shipped in DD 2.0.3.0-mac-edge.
